### PR TITLE
Highlight subject line trailing period

### PR DESCRIPTION
**Pull request by walles**
Because http://chris.beams.io/posts/git-commit/#seven-rules says: "Do
not end the subject line with a period".

This change comes with unit tests for the subject line highlighting
and as such most likely fixes issues with the previous highlighting
as well.

### CODE REVIEW

**Comment by thomasjo** [comment]
Thanks for this!

Personally, I really, really dislike using (complicated) logic in test assertions, since (a) tests typically because much harder to reason about, and (b) it makes it much more likely to report both false positive and false negatives due to bugs in the logic.

In other words, I disapprove of the usage of the introduced `hilight` method. At least how it's being used. Any usage of `expect` should be trivial to reason about. It should come down to comparing simple values that does not require mentally parsing some other function.

In summary, I support the intended change in this PR, but not how the specs have been implemented — they need to be simplified.

/cc @atom/feedback

**Comment by walles** [comment]
So if we take one of the test cases as an example:
```coffeescript
expect(
  hilight("h2345678901234567890123456789012345678901234567890123456789012345678.")).
  toEqual("i                                                 ddddddddddddddddddi")
```

What would you consider a better / more readable / easier to reason about way to express that:
* The first char should be marked as invalid
* Chars 2-50 shouldn't be marked any particular way
* Chars 51-68 should be marked as deprecated
* Char 69 should be marked as illegal

Since both lines are aligned underneath each other I think this way is pretty straight forward, but that's kind of a given since I wrote it :).

But some hints on what you'd consider a better way to formulate these 24 test cases would be nice!

**Comment by 50Wliu** [comment]
The equivalent verbose way would be this, right?
```coffee
expect(tokens[0]).toEqual value: 'h', scopes: ['invalid scopes']
expect(tokens[1]).toEqual value: '2345678901234567890123456789012345678901234567890', scopes: ['normal scopes']
expect(tokens[2]).toEqual value: '123456789012345678', scopes: ['deprecated scopes']
expect(tokens[3]).toEqual value: '.', scopes: ['invalid scopes']
```
That's still very readable, and often-times you can just test for specific tokens, eg the last one in this case if the other ones have their own spec coverage.

**Comment by walles** [comment]
@50Wliu, I just tried your suggestion and it didn't work for me.

There are more than just one scope per token, and adding them all makes everything quite unreadable IMO.

Added documentation for the `hilight()` method; that was definitely needed.

**Comment by thomasjo** [comment]
@walles Can you elaborate on exactly what didn't work for you?

I'm still opposed to allowing this `hilight` function. We should not have "advanced" logic in tests. Tests should always be explicit. Using a function like the one implemented can have wildly unexpected consequences — it can trivially hide true negatives in tests if the function implementation contains some non-trivial bug.

**TL;DR** It's a problem when tests need tests to verify their validity.

**Comment by walles** [comment]
I updated this method according to @50Wliu's sample code:
```coffeescript
    it "highlights subject lines of 69 chars correctly", ->
      {tokens} = grammar.tokenizeLine("123456789012345678901234567890123456789012345678901234567890123456789", null, true)
      expect(tokens[0]).toEqual value: '12345678901234567890123456789012345678901234567890', scopes: ['normal scopes']
      expect(tokens[1]).toEqual value: '1234567890123456789', scopes: ['deprecated scopes']

      {tokens} = grammar.tokenizeLine("g23456789012345678901234567890123456789012345678901234567890123456789", null, true)
      expect(tokens[0]).toEqual value: 'g', scopes: ['invalid scopes']
      expect(tokens[1]).toEqual value: '2345678901234567890123456789012345678901234567890', scopes: ['normal scopes']
      expect(tokens[2]).toEqual value: '123456789012345679', scopes: ['deprecated scopes']

      {tokens} = grammar.tokenizeLine("12345678901234567890123456789012345678901234567890123456789012345678.", null, true)
      expect(tokens[0]).toEqual value: '12345678901234567890123456789012345678901234567890', scopes: ['normal scopes']
      expect(tokens[1]).toEqual value: '123456789012345678', scopes: ['deprecated scopes']
      expect(tokens[2]).toEqual value: '.', scopes: ['invalid scopes']

      {tokens} = grammar.tokenizeLine("h2345678901234567890123456789012345678901234567890123456789012345678.", null, true)
      expect(tokens[0]).toEqual value: 'h', scopes: ['invalid scopes']
      expect(tokens[1]).toEqual value: '2345678901234567890123456789012345678901234567890', scopes: ['normal scopes']
      expect(tokens[2]).toEqual value: '123456789012345678', scopes: ['deprecated scopes']
      expect(tokens[3]).toEqual value: '.', scopes: ['invalid scopes']
```

And here's the output:
```
~/s/atom-language-git (walles/highlight-period|✚1) $ apm test
.....F.................

Git grammars
  Git commit messages
    it highlights subject lines of 69 chars correctly
      Expected { value : '12345678901234567890123456789012345678901234567890', scopes : [ 'text.git-commit', 'meta.scope.message.git-commit' ] } to equal { value : '12345678901234567890123456789012345678901234567890', scopes : [ 'normal scopes' ] }.
        at .<anonymous> (/Users/johan/src/atom-language-git/spec/git-spec.coffee:101:25)
      Expected { value : '123456789012345678', scopes : [ 'text.git-commit', 'meta.scope.message.git-commit', 'invalid.deprecated.line-too-long.git-commit' ] } to equal { value : '1234567890123456789', scopes : [ 'deprecated scopes' ] }.
        at .<anonymous> (/Users/johan/src/atom-language-git/spec/git-spec.coffee:102:25)
      Expected { value : 'g', scopes : [ 'text.git-commit', 'meta.scope.message.git-commit', 'invalid.illegal.first-char-should-be-uppercase.git-commit' ] } to equal { value : 'g', scopes : [ 'invalid scopes' ] }.
        at .<anonymous> (/Users/johan/src/atom-language-git/spec/git-spec.coffee:105:25)
      Expected { value : '2345678901234567890123456789012345678901234567890', scopes : [ 'text.git-commit', 'meta.scope.message.git-commit' ] } to equal { value : '2345678901234567890123456789012345678901234567890', scopes : [ 'normal scopes' ] }.
        at .<anonymous> (/Users/johan/src/atom-language-git/spec/git-spec.coffee:106:25)
      Expected { value : '123456789012345678', scopes : [ 'text.git-commit', 'meta.scope.message.git-commit', 'invalid.deprecated.line-too-long.git-commit' ] } to equal { value : '123456789012345679', scopes : [ 'deprecated scopes' ] }.
        at .<anonymous> (/Users/johan/src/atom-language-git/spec/git-spec.coffee:107:25)
      Expected { value : '12345678901234567890123456789012345678901234567890', scopes : [ 'text.git-commit', 'meta.scope.message.git-commit' ] } to equal { value : '12345678901234567890123456789012345678901234567890', scopes : [ 'normal scopes' ] }.
        at .<anonymous> (/Users/johan/src/atom-language-git/spec/git-spec.coffee:110:25)
      Expected { value : '123456789012345678', scopes : [ 'text.git-commit', 'meta.scope.message.git-commit', 'invalid.deprecated.line-too-long.git-commit' ] } to equal { value : '123456789012345678', scopes : [ 'deprecated scopes' ] }.
        at .<anonymous> (/Users/johan/src/atom-language-git/spec/git-spec.coffee:111:25)
      Expected { value : '.', scopes : [ 'text.git-commit', 'meta.scope.message.git-commit', 'invalid.illegal.subject-no-trailing-period.git-commit' ] } to equal { value : '.', scopes : [ 'invalid scopes' ] }.
        at .<anonymous> (/Users/johan/src/atom-language-git/spec/git-spec.coffee:112:25)
      Expected { value : 'h', scopes : [ 'text.git-commit', 'meta.scope.message.git-commit', 'invalid.illegal.first-char-should-be-uppercase.git-commit' ] } to equal { value : 'h', scopes : [ 'invalid scopes' ] }.
        at .<anonymous> (/Users/johan/src/atom-language-git/spec/git-spec.coffee:115:25)
      Expected { value : '2345678901234567890123456789012345678901234567890', scopes : [ 'text.git-commit', 'meta.scope.message.git-commit' ] } to equal { value : '2345678901234567890123456789012345678901234567890', scopes : [ 'normal scopes' ] }.
        at .<anonymous> (/Users/johan/src/atom-language-git/spec/git-spec.coffee:116:25)
      Expected { value : '123456789012345678', scopes : [ 'text.git-commit', 'meta.scope.message.git-commit', 'invalid.deprecated.line-too-long.git-commit' ] } to equal { value : '123456789012345678', scopes : [ 'deprecated scopes' ] }.
        at .<anonymous> (/Users/johan/src/atom-language-git/spec/git-spec.coffee:117:25)
      Expected { value : '.', scopes : [ 'text.git-commit', 'meta.scope.message.git-commit', 'invalid.illegal.subject-no-trailing-period.git-commit' ] } to equal { value : '.', scopes : [ 'invalid scopes' ] }.
        at .<anonymous> (/Users/johan/src/atom-language-git/spec/git-spec.coffee:118:25)


Finished in 0.613 seconds
23 tests, 103 assertions, 12 failures, 0 skipped

Tests failed
~/s/atom-language-git (walles/highlight-period|✚1) $
```

Not sure why this worked for @50Wliu, but for me it fails.

A working example of how to rewrite the tests to your liking would be much appreciated!

**Comment by thomasjo** [comment]
His example code was exactly that, an example. Where he put `invalid scopes` and `deprecated scopes` you simply need to define the ordered set of expected scopes. Similarly, you'll need to specify the token position to the "important" parts of the string. E.g. if you want to test for an invalid scope due to a subject ending with a period, then check e.g. that the last token (using a hardcoded index) matches what you expect to find.

As a more complete example for where you are testing `xxx`, the first assertion might perhaps be
``` coffee
expect(tokens[0]).toEqual value: 'h', scopes: ['text.git-commit', 'meta.scope.message.git-commit', 'invalid.illegal.first-char-should-be-uppercase.git-commit']
```

That assertion verifies that a subject line starting with a lowercase `h` should be flagged as invalid, with those explicit scope names.

Also, when you are testing a specific thing such as the above, it's much "friendlier" if you construct simple examples that produce the expected outcomes --- I would change the string you are tokenizing to something like
``` coffee
{tokens} = grammar.tokenizeLine("foo", null, true)
```

Hopefully this answers your questions.

**Comment by 50Wliu** [comment]
The `, null, true` isn't needed by the way.

**Comment by thomasjo** [comment]
@50Wliu Thanks for pointing that out. I edited my simple example, and now it's even more readable ✨

**Comment by walles** [comment]
@50Wliu, @thomasjo, the `, null, true` indicates that these tests are for the subject line:
https://atom.io/docs/api/v1.12.6/Grammar#instance-tokenizeLine

Without that the tests will test the wrong thing.

File: grammars/git commit message.cson [hunk 1]
```diff
@@ -23,26 +23,37 @@
...
+          '5':
+            'name': 'invalid.illegal.subject-no-trailing-period.git-commit'
+          '6':
+            'name': 'invalid.deprecated.line-too-long.git-commit'
       }
       {
-        # Subject line 51-69 chars
-        'match': '\\A(?!#).{50}(.{0,19})'
+        # Subject line 70 chars or longer
+        'match': '\\A(?!#)(([a-z])|.).{49}(.{0,19})(.+)$'
```
Comment (50Wliu) [inline grammars/git commit message.cson]: This also needs to test for a trailing period.

File: spec/git-spec.coffee [hunk 2]
```diff
@@ -14,13 +14,147 @@ describe "Git grammars", ->
       expect(grammar.scopeName).toBe "source.git-config"
 
   describe "Git commit messages", ->
+    scopeNormal =  [ 'text.git-commit', 'meta.scope.message.git-commit' ]
```
Comment (50Wliu) [inline spec/git-spec.coffee]: Nit: remove the padded whitespace in these brackets.
`['text.git-commit', 'meta.scope.message.git-commit']`

File: spec/git-spec.coffee [hunk 3]
```diff
@@ -14,13 +14,147 @@ describe "Git grammars", ->
...
+
+    scopeLeadingLowercase =
+      [ 'text.git-commit', 'meta.scope.message.git-commit', 'invalid.illegal.first-char-should-be-uppercase.git-commit' ]
+
+    scopeTrailingPeriod =
+      [ 'text.git-commit', 'meta.scope.message.git-commit', 'invalid.illegal.subject-no-trailing-period.git-commit' ]
+
+    scopeTooLong1 = [ 'text.git-commit', 'meta.scope.message.git-commit', 'invalid.deprecated.line-too-long.git-commit' ]
```
Comment (50Wliu) [inline spec/git-spec.coffee]: How about scopeLineOver50 and scopeLineOver70?

**Comment by 50Wliu** [comment]
Looks good.  Just a few minor things.

File: spec/git-spec.coffee (hunk 2 above)
Comment (walles) [inline spec/git-spec.coffee]: Are there coffeescript linter rules for this?

File: grammars/git commit message.cson (hunk 1 above)
Comment (walles) [inline grammars/git commit message.cson]: Depends I guess; is a post-70-chars trailing period mostly an illegal period or mostly a too long line?

Comment (50Wliu) [inline grammars/git commit message.cson]: It can be both, can't it? :wink:
Something like the following should work:
`\\A(?!#)(([a-z])|.).{49}(.{0,19})(.+?(\\.)?)$`

Comment (walles) [inline grammars/git commit message.cson]: I don't think it can actually be both; my question was really related to what `name:` should be.

Right now the `name:` for a post-70-chars trailing period would be `invalid.illegal.line-too-long.git-commit` but not `invalid.illegal.subject-no-trailing-period.git-commit`.

With your suggested change the `name:` for a post 70-chars-trailing period would be `invalid.illegal.subject-no-trailing-period.git-commit` but not `invalid.illegal.line-too-long.git-commit`.

Or can `name:` take an array? The only documentation I managed to find doesn't seem to think so, but maybe it is possible? Do you have a better documentation link (as in "for Atom, not for TextMate")?
https://manual.macromates.com/en/language_grammars#rule_keys

So if we can't have both, which is more important for a post-70-chars trailiing period, `invalid.illegal.subject-no-trailing-period.git-commit` or `invalid.illegal.line-too-long.git-commit`?

Comment (50Wliu) [inline grammars/git commit message.cson]: I would say the line too long takes precedence, but also tokenize the period as invalid separately from the long line.

File: grammars/git commit message.cson [hunk 4]
```diff
@@ -23,26 +23,39 @@
...
+          '5':
+            'name': 'invalid.illegal.subject-no-trailing-period.git-commit'
+          '6':
+            'name': 'invalid.deprecated.line-too-long.git-commit'
       }
       {
-        # Subject line 51-69 chars
-        'match': '\\A(?!#).{50}(.{0,19})'
+        # Subject line 70 chars or longer
+        'match': '\\A(?!#)(([a-z])|.).{49}(.{0,19})(.*?)([.]?)$'
```
Comment (50Wliu) [inline grammars/git commit message.cson]: Why is this not `\\.?`?

Comment (walles) [inline grammars/git commit message.cson]: Because I find it more readable not have to escape things.

But I've pushed a new commit which changes it to `\\.` now.

**Comment by 50Wliu** [comment]
@thomasjo ✅ ?

**Comment by 50Wliu** [comment]
@walles want to tackle that real quick?  Explicitly saying how the lines will be tokenized (eg valid, too long, etc) would probably suffice.

**Comment by walles** [comment]
@50Wliu, given that this PR is now ✅ by all parties and that it took 29 comments and two weeks to get here, I'd prefer to have this merged as-is.

**Comment by walles** [comment]
@50Wliu, @thomasjo, thanks for staying with this and (re)reviewing with me until merge :)

**Comment by thomasjo** [comment]
@walles Big thanks to you for your patience and perseverance! 🙇
